src/lib.rs

```diff
@@ -310,71 +309,41 @@ impl Watch {
+
+        let handler = WatchEventHandler {
+            watch: self.clone(),
+            command,
+            child,
+            command_start: Instant::now(),
+        };
+
+        let mut watcher =
+            notify::recommended_watcher(handler).context("could not initialize watcher")?;
 
         for path in &self.watch_paths {
-            match watcher.watch(&path, RecursiveMode::Recursive) {
+            match watcher.watch(path, RecursiveMode::Recursive) {
                 Ok(()) => log::trace!("Watching {}", path.display()),
                 Err(err) => log::error!("cannot watch {}: {}", path.display(), err),
             }
         }
 
-        let mut child = command.spawn().context("cannot spawn command")?;
-        let mut command_start = Instant::now();
-
-        loop {
-            match rx.recv() {
-                Ok(notify::RawEvent {
-                    path: Some(path),
-                    op: Ok(op),
-                    ..
-                }) if !self.is_excluded_path(&path)
-                    && path.exists()
-                    && !self.is_hidden_path(&path)
-                    && !self.is_backup_file(&path)
-                    && op != notify::Op::CREATE
-                    && op != notify::Op::RENAME
-                    && command_start.elapsed() >= self.debounce =>
-                {
-                    log::trace!("Detected changes at {} | {:?}", path.display(), op);
-                    #[cfg(unix)]
-                    {
-                        let now = Instant::now();
-
-                        unsafe {
-                            log::trace!("Killing watch's command process");
-                            libc::kill(
-                                child.id().try_into().expect("cannot get process id"),
-                                libc::SIGTERM,
-                            );
-                        }
-
-                        while now.elapsed().as_secs() < 2 {
-                            std::thread::sleep(Duration::from_millis(200));
-                            if let Ok(Some(_)) = child.try_wait() {
-                                break;
-                            }
-                        }
-                    }
-
-                    match child.try_wait() {
-                        Ok(Some(_)) => {}
-                        _ => {
-                            let _ = child.kill();
-                            let _ = child.wait();
-                        }
-                    }
+        Ok(())
+    }
 
-                    log::info!("Re-running command");
-                    child = command.spawn().context("cannot spawn command")?;
-                    command_start = Instant::now();
-                }
-                Ok(event) => log::trace!("Ignoring changes in {:?}", event),
-                Err(err) => log::error!("watch error: {}", err),
-            }
-        }
+    fn compare_event(&self, event: &notify::Event, command_duration: Duration) -> bool {
```

Comment:
I think it would make more sense to pass the instant when the command has been started instead of a duration like that. And call `.elapsed()` in this function instead. I guess that's a matter of taster